product_customer_code/models/product_customer_code.py

```diff
@@ -0,0 +1,35 @@
+# coding: utf-8
+
+from openerp import api, fields, models
+
+
+class ProductCustomerCode(models.Model):
+
+    _name = "product.customer.code"
+    _description = "Add manies Code of Customer's"
+
+    _rec_name = 'product_code'
+
+    product_code = fields.Char('Customer Product Code', required=True,
+                               index=True, help="""This customer's product
+                               code will be used when searching into a request
+                               for quotation.""")
+    product_name = fields.Char('Customer Product Name', help="""This customer's
+                               product name will be used when searching into a
+                               request for quotation.""")
+    product_id = fields.Many2one('product.product', 'Product', required=True,
+                                 help='Product Identification')
+    partner_id = fields.Many2one('res.partner', 'Customer', required=True,
+                                 index=True, help='Partner Reference')
```

Comment:
You are using a search with the following domain: `('product_code', '=', name), ('partner_id', '=', partner_id)`
So you don't need this `index=True` you will need a database complex index with product_code, partner_id